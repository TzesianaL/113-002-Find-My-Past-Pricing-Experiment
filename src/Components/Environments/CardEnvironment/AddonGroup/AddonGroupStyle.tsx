import styled from 'styled-components'

/**
 * Styled components are used in order to style AddonGroup component. Props are passed through ThemeProvider.
 */
export const AddonGroupStyle = styled.div`
  display: flex;
  flex-direction: column;
  width: 100%;
`
