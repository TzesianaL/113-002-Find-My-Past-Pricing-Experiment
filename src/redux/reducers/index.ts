import { combineReducers } from 'redux'
import { dataStoreReducer } from './dataStore/dataStore'
import { outputStoreReducer } from './output/outputSlice'
import { logicStoreReducer } from './logicStore/logicStore'
import { pageDurationStoreReducer } from './pageDurationStore/pageDurationStore'
import { actionLogStoreReducer } from './actionLogStore/actionLogStore'

const rootReducer = combineReducers({
  data: dataStoreReducer,
  logic: logicStoreReducer,
  output: outputStoreReducer,
  pageDuration: pageDurationStoreReducer,
  actionLog: actionLogStoreReducer,
})

export default rootReducer
export type RootState = ReturnType<typeof rootReducer>
