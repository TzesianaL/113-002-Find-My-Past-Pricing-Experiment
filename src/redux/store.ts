import { configureStore, EnhancedStore } from '@reduxjs/toolkit'
import merge from 'lodash-es/merge'
import rootReducer, { RootState } from './reducers'
import { initialState as dataInitialState } from './reducers/dataStore/dataStore'
import { initialState as actionLogInitialState } from './reducers/actionLogStore/actionLogStore'
import { initialState as logicInitialState } from './reducers/logicStore/logicStore'
import { initialState as outputInitialState } from './reducers/output/outputSlice'
import { initialState as pageDurationInitialState } from './reducers/pageDurationStore/pageDurationStore'

export const store = configureStore({
  reducer: rootReducer,
})

const defaultStoreState: RootState = {
  data: dataInitialState,
  logic: logicInitialState,
  actionLog: actionLogInitialState,
  output: outputInitialState,
  pageDuration: pageDurationInitialState,
}

// eslint-disable-next-line @typescript-eslint/ban-types
type DeepPartial<T> = T extends object
  ? {
      [P in keyof T]?: DeepPartial<T[P]>
    }
  : T

export const preloadedStore = (
  state: DeepPartial<RootState>
): EnhancedStore<RootState> =>
  configureStore({
    reducer: rootReducer,
    preloadedState: merge({}, defaultStoreState, state),
  })
