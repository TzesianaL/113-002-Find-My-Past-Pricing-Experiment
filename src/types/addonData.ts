import { Periods } from './periods'

export interface AddonData {
  id: string
  title: string
  subtitle?: string
  features?: string[]
  price?: number
  pricePeriod?: Periods
  includedProductIDs?: string[]
  excludedProductIDs?: string[]
  mutex?: boolean
  group?: number
  mutexGroup4?: boolean
  mutexGroup2?: boolean
}
