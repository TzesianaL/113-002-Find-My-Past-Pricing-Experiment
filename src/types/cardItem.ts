import { Periods } from './periods'

interface BasketCardOptions {
  prePriceText?: string
  postPriceText?: string
}

export type ContractDuration = '24 months' | '30 days'

// Type guard function for runtime validation of ContractDuration
export const isContractDuration = (
  value: string
): value is ContractDuration => {
  return value === '24 months' || value === '30 days'
}

export interface CardItem {
  id: string
  title: string
  subtitle: string
  subtitle2?: string
  price: number
  currency?: string
  priceString?: string
  features: string[]
  cardColor?: string
  highlightedText?: string
  headerImage?: string
  upfrontPrice?: number
  basketOptions?: BasketCardOptions
  pricePeriod?: Periods
  contractDuration?: ContractDuration
  parentSpeedId?: string
}

export interface SpeedCardItem {
  title: string
  subtitle: string
  price: number
  currency?: string
  highlightedText?: string
  id: string
  pricePeriod?: Periods
  purchaseButtonText?: string
  mainSpeed: string
  downloadSpeed: string
  downloadExample: string
  downloadTime: string
  uploadSpeed: string
  uploadExample: string
  uploadTime: string
  speed?: number
}
