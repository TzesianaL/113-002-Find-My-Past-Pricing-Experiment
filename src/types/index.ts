export * from './addonData'
export * from './cardItem'
export * from './layout'
export * from './periods'
export * from './reduxData'
export * from './splashItem'
export * from './pcw'
export * from './environmentType'
