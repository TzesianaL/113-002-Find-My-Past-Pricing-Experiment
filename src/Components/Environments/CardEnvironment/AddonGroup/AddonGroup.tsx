import React, { FC } from 'react'
import Addon from 'Components/Environments/CardEnvironment/Addon/Addon'
import { BlockLayout, AddonData } from 'types'
import { AddonGroupStyle } from './AddonGroupStyle'

export interface AddonGroupProps {
  /**
   * Items contains all the add on and its features
   */
  layout: BlockLayout
  data: AddonData[]
}

/**
 * AddonsGroup is used to wrap all components we render in order to create the list of add on we display on add on page .
 * ThemeProvider component is used. By using the theme object we provide same styling options to all components.
 */
const AddonGroup: FC<AddonGroupProps> = ({ layout, data }) => {
  return (
    <AddonGroupStyle>
      {data.map((item, i) => (
        <Addon
          // disabling the error since we are sure that the add on items won't change
          // but this is better replaced by unique ids
          // eslint-disable-next-line react/no-array-index-key
          key={i}
          data={item}
          layout={layout}
        />
      ))}
    </AddonGroupStyle>
  )
}

export default AddonGroup
