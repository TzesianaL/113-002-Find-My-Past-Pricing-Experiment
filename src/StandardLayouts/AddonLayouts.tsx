import { BlockLayout, BlockEnums } from 'types'

export const horizontalThreeColumnAddonLayout: BlockLayout = {
  blockType: BlockEnums.HStack,
  blocks: [
    {
      blockType: BlockEnums.VStack,
      blocks: [
        { blockType: BlockEnums.Title },
        {
          blockType: BlockEnums.Subtitle,
          styleOverride: {
            color: 'gray',
          },
        },
      ],
      styleOverride: {
        flex: '0 0 20em',
        alignSelf: 'flex-start',
        alignItems: 'center',
      },
    },
    {
      blockType: BlockEnums.Features,
      styleOverride: {
        flex: '1 0 15em',
        maxWidth: '20em',
      },
    },
    {
      blockType: BlockEnums.Price,
      styleOverride: { flex: '0 0 auto', margin: '0 2em' },
    },
    {
      blockType: BlockEnums.AddRemoveButton,
      styleOverride: {
        flex: '0 0 7em',
        margin: '0 2em',
      },
    },
  ],
}

export const cardAddonLayout: BlockLayout = {
  blockType: BlockEnums.VStack,
  blocks: [
    {
      blockType: BlockEnums.Title,
      styleOverride: {
        color: 'black',
        alignSelf: 'left',
        textAlign: 'left',
      },
    },
    {
      blockType: BlockEnums.Subtitle,
      styleOverride: {
        color: 'black',
        alignSelf: 'left',
        textAlign: 'left',
      },
    },
    {
      blockType: BlockEnums.Features,
      styleOverride: {
        color: 'black',
      },
    },
    {
      blockType: BlockEnums.Price,
      styleOverride: {
        color: 'black',
        alignSelf: 'left',
        fontWeight: 'bold',
      },
    },
    {
      blockType: BlockEnums.AddRemoveButton,
      styleOverride: {
        alignSelf: 'center',
        width: '300px',
        background: 'black',
        color: 'white',
      },
    },
  ],
  styleOverride: {
    width: '18em',
    height: '15em',
    border: '2px solid black',
    padding: '1em',
    background: 'white',
    justifyContent: 'space-around',
    margin: '1em',
    color: 'black',
    borderRadius: '5px',
  },
}

export const WifiMaxAddonLayout: BlockLayout = {
  blockType: BlockEnums.VStack,
  blocks: [
    {
      blockType: BlockEnums.Title,
      styleOverride: {
        color: 'black',
        alignSelf: 'left',
        textAlign: 'left',
      },
    },
    {
      blockType: BlockEnums.Subtitle,
      styleOverride: {
        color: 'black',
        alignSelf: 'left',
        textAlign: 'left',
      },
    },
    {
      blockType: BlockEnums.Features,
      styleOverride: {
        color: 'black',
      },
    },
    {
      blockType: BlockEnums.Price,
      styleOverride: {
        color: 'black',
        alignSelf: 'left',
        fontWeight: 'bold',
      },
    },
    {
      blockType: BlockEnums.AddRemoveButton,
      styleOverride: {
        alignSelf: 'center',
        width: '300px',
        background: 'black',
        color: 'white',
      },
    },
  ],
  styleOverride: {
    width: '18em',
    height: '16em',
    border: '2px solid black',
    padding: '1em',
    background: 'white',
    justifyContent: 'space-around',
    margin: '1em',
    color: 'black',
    borderRadius: '5px',
  },
}
export const highlightedCardAddonLayout: BlockLayout = {
  blockType: BlockEnums.VStack,
  blocks: [
    {
      blockType: BlockEnums.Title,
      styleOverride: {
        color: 'black',
      },
    },
    {
      blockType: BlockEnums.Subtitle,
      styleOverride: {
        color: 'black',
        alignSelf: 'center',
      },
    },
    {
      blockType: BlockEnums.Features,
      styleOverride: {
        color: 'black',
      },
    },
    {
      blockType: BlockEnums.Price,
      styleOverride: {
        color: 'black',
        alignSelf: 'center',
        fontWeight: 'bold',
      },
    },
    {
      blockType: BlockEnums.AddRemoveButton,
      styleOverride: {
        alignSelf: 'center',
        width: '200px',
        background: 'black',
        color: 'white',
      },
    },
  ],
  styleOverride: {
    width: '18em',
    height: '20em',
    border: '2px solid black',
    padding: '1em',
    background: 'white',
    justifyContent: 'space-around',
    margin: '1em',
    color: 'black',
    borderRadius: '5px',
  },
}
export const testLayout: BlockLayout = {
  blockType: BlockEnums.HStack,
  blocks: [
    {
      blockType: BlockEnums.VStack,
      blocks: [
        {
          blockType: BlockEnums.Title,
          styleOverride: {
            color: 'red',
          },
        },
        {
          blockType: BlockEnums.Subtitle,
          styleOverride: {
            color: 'gray',
          },
        },
      ],
      styleOverride: {
        flex: '0 0 50%',
        alignItems: 'center',
      },
    },
    {
      blockType: BlockEnums.Features,
      styleOverride: {
        flex: '1 0 auto',
        width: 'fit-content',
        alignItems: 'flex-start',
      },
    },
  ],
}

export const StorePageAddonLayout: BlockLayout = {
  blockType: BlockEnums.VStack,
  blocks: [
    {
      blockType: BlockEnums.Title,
      styleOverride: {
        color: 'black',
      },
    },
    {
      blockType: BlockEnums.Subtitle,
      styleOverride: {
        color: 'black',
        alignSelf: 'center',
      },
    },
    {
      blockType: BlockEnums.Features,
      styleOverride: {
        color: 'black',
      },
    },
    {
      blockType: BlockEnums.Price,
      styleOverride: {
        color: 'black',
        alignSelf: 'center',
        fontWeight: 'bold',
      },
    },
    {
      blockType: BlockEnums.AddRemoveButton,
      styleOverride: {
        alignSelf: 'center',
        width: '200px',
        background: 'black',
        color: 'white',
      },
    },
  ],
  styleOverride: {
    width: '20em',
    height: '28em',
    border: '2px solid black',
    padding: '1em',
    background: 'white',
    justifyContent: 'space-around',
    margin: '1em',
    color: 'black',
    borderRadius: '7px',
  },
}

export const StorePageAddonLayoutScenarios123: BlockLayout = {
  blockType: BlockEnums.VStack,
  blocks: [
    {
      blockType: BlockEnums.Title,
      styleOverride: {
        color: 'black',
      },
    },
    {
      blockType: BlockEnums.Subtitle,
      styleOverride: {
        color: 'black',
        alignSelf: 'center',
        textAlign: 'center',
      },
    },
    {
      blockType: BlockEnums.Features,
      styleOverride: {
        color: 'black',
        textAlign: 'center',
      },
    },
    {
      blockType: BlockEnums.Price,
      styleOverride: {
        color: 'black',
        alignSelf: 'center',
        fontWeight: 'bold',
      },
    },
    {
      blockType: BlockEnums.AddRemoveButton,
      styleOverride: {
        alignSelf: 'center',
        width: '200px',
        background: 'black',
        color: 'white',
      },
    },
  ],
  styleOverride: {
    width: '20em',
    height: '31em',
    border: '2px solid #83358e',
    padding: '1em',
    background: 'white',
    justifyContent: 'space-around',
    margin: '1em',
    color: 'black',
    borderRadius: '7px',
  },
}

export const xtraBroadBandLayout: BlockLayout = {
  blockType: BlockEnums.VStack,
  blocks: [
    {
      blockType: BlockEnums.Title,
      styleOverride: {
        color: 'black',
      },
    },
    {
      blockType: BlockEnums.Subtitle,
      styleOverride: {
        color: 'black',
        alignSelf: 'center',
        textAlign: 'center',
      },
    },
    {
      blockType: BlockEnums.Features,
      styleOverride: {
        color: 'black',
        textAlign: 'center',
      },
    },
    {
      blockType: BlockEnums.Price,
      styleOverride: {
        color: 'black',
        alignSelf: 'center',
        fontWeight: 'bold',
      },
    },
    {
      blockType: BlockEnums.AddRemoveButton,
      styleOverride: {
        alignSelf: 'center',
        width: '200px',
        background: 'black',
        color: 'white',
      },
    },
  ],
  styleOverride: {
    width: '18em',
    height: '25em',
    border: '2px solid #228B22',
    padding: '1em',
    background: '#D3F8D3',
    justifyContent: 'space-around',
    margin: '1em',
    color: 'black',
    borderRadius: '7px',
  },
}
