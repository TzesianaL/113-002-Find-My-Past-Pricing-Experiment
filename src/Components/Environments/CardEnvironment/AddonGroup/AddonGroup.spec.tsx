import React from 'react'
import 'jest-styled-components'
import { render } from '@testing-library/react'
import { horizontalThreeColumnAddonLayout } from 'StandardLayouts/AddonLayouts'
import { AddonData } from 'types'
import { Provider } from 'react-redux'
import { store } from 'redux/store'
import AddonGroup from './AddonGroup'

describe('<AddonGroup/>', () => {
  const Addons: AddonData[] = [
    {
      id: 'Addon_1',
      title: 'Basic',
      price: 9.99,
      features: ['Feature 1'],
    },
    {
      id: 'Addon_2',
      title: 'Better',
      price: 29.99,
      features: ['Feature 1', 'Feature 2'],
    },
    {
      id: 'Addon_3',
      title: 'Premium',
      price: 39.99,
      features: ['Feature 1', 'Feature 2', 'Feature 3', 'Feature 4'],
    },
  ]
  it('should render AddonGroup correctly', () => {
    const { asFragment } = render(
      <Provider store={store}>
        <AddonGroup data={Addons} layout={horizontalThreeColumnAddonLayout} />
      </Provider>
    )

    expect(asFragment()).toMatchSnapshot()
  })
})
