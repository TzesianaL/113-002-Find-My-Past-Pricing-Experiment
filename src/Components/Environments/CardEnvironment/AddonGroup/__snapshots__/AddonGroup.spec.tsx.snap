// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddonGroup/> should render AddonGroup correctly 1`] = `
<DocumentFragment>
  .c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 img {
  margin: 1rem;
  width: 160px;
  border-radius: 10px;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  text-align: center;
  text-transform: uppercase;
  font-weight: bold;
  font-family: BoldVf !important;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  font-size: 1.2em;
  margin: 0.5em 0em 0.5em 0em;
}

.c5 {
  color: black;
  font-size: em;
  padding-left: 2em;
  padding-right: 2em;
  width: calc(100% - 4em);
}

.c5 li {
  padding-left: 0.25em;
  margin-bottom: 1.5em;
}

.c5 li::marker {
  content: '✓';
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  font-size: 1rem;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.c7 {
  color: #ffffff;
  background-color: #003D59;
  cursor: pointer;
  font-size: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  border: 0px;
  border-radius: 6px;
  padding: 0.7em 1em;
}

.c7:active {
  background-color: #336D89;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  font-size: 1.2rem;
  font-family: BoldVf;
  margin: 0em 0em 0.5em 0em;
}

.c6 del {
  color: red;
}

.c6 .pricePeriodPlaceHolder {
  font-size: small;
  color: grey;
  line-height: 23px;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
}

<div
    class="c0"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
          style="flex: 0 0 20em; align-self: flex-start; align-items: center;"
        >
          <div
            class="c4"
          >
            Basic
          </div>
        </div>
        <ul
          class="c5"
          style="flex: 1 0 15em; max-width: 20em;"
        >
          <div>
            Feature 1
          </div>
        </ul>
        <div
          class="c6"
          style="flex: 0 0 auto; margin: 0px 2em;"
        >
          <span
            class="pricePeriodPlaceHolder"
          >
             
          </span>
            
          <strong>
            £9.99
          </strong>
             
          <span
            class="pricePeriodPlaceHolder"
          >
            a month
          </span>
           
        </div>
        <button
          class="c7"
          style="flex: 0 0 7em; margin: 0px 2em;"
        >
          Add
        </button>
      </div>
    </div>
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
          style="flex: 0 0 20em; align-self: flex-start; align-items: center;"
        >
          <div
            class="c4"
          >
            Better
          </div>
        </div>
        <ul
          class="c5"
          style="flex: 1 0 15em; max-width: 20em;"
        >
          <div>
            Feature 1
          </div>
          <div>
            Feature 2
          </div>
        </ul>
        <div
          class="c6"
          style="flex: 0 0 auto; margin: 0px 2em;"
        >
          <span
            class="pricePeriodPlaceHolder"
          >
             
          </span>
            
          <strong>
            £29.99
          </strong>
             
          <span
            class="pricePeriodPlaceHolder"
          >
            a month
          </span>
           
        </div>
        <button
          class="c7"
          style="flex: 0 0 7em; margin: 0px 2em;"
        >
          Add
        </button>
      </div>
    </div>
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
          style="flex: 0 0 20em; align-self: flex-start; align-items: center;"
        >
          <div
            class="c4"
          >
            Premium
          </div>
        </div>
        <ul
          class="c5"
          style="flex: 1 0 15em; max-width: 20em;"
        >
          <div>
            Feature 1
          </div>
          <div>
            Feature 2
          </div>
          <div>
            Feature 3
          </div>
          <div>
            Feature 4
          </div>
        </ul>
        <div
          class="c6"
          style="flex: 0 0 auto; margin: 0px 2em;"
        >
          <span
            class="pricePeriodPlaceHolder"
          >
             
          </span>
            
          <strong>
            £39.99
          </strong>
             
          <span
            class="pricePeriodPlaceHolder"
          >
            a month
          </span>
           
        </div>
        <button
          class="c7"
          style="flex: 0 0 7em; margin: 0px 2em;"
        >
          Add
        </button>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
